drivers/clock_controller: misc. fixes for stm32WB and stm32WL

str4t0m opened the pull request:
In specific configurations the stm32 clock control driver has a few issues. which this pr intends to solve.

- stm32wb and stm32wl: fix struct wrongly de-referenced(in case msi is selected as system clock)
- stm32wl: restructure cpu2 prescaler assignment
- stm32wb: only has a single msi range: therefore hal macros differ. This must be taken account for to avoid build errors.
- stm32wb and stm32wl: fix missing initialization of ahb prescalers in case HSI or HSE are selected as system clock.
- stm32wb and stm32wl: fix flash latency calculation in the initialization function by using the correct prescaler
  (it is not the same as cpu1 prescaler).

str4t0m committed:
drivers/clock_control: stm32 fix struct wrongly dereferenced

This commit fixes several occurences where a struct members is wrongly
dereferenced, which causes a compile error in case the msi clock is used
as system clock.
Only affects stm32wb and stm32wl with MSI selected as sysclock.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/clock_control: stm32 wb, wl: fix missing init of rcc prescalers

This commit adds missing initialization of rcc prescalers for
stm32wb and stm32wl series when hse or hsi are selected as
system clock.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/clock: stm32wl config restructure cpu2 prescaler assignment

This commmit restructures the clock_controller code such that the cpu2
prescler assignment later can be excluded for single core socs.

The stm32wl mcu line has variants with a single cortex-m4 core
(stm32wle5), therefore the prescaler for the second clock should
only be set for dual core socs.
This commit still checks for the complete series
(CONFIG_SOC_SERIES_STM32WLX) as the single core variants are not
yet introduced, later the condition should check for a flag like
CONFIG_SOC_STM32WL5X instead.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/clock_control: fix: stm32wb series has only a single msi range

stm32wb only has a single msi clock range, in contrast to wl, l4, l5
which have a second range that is active after exit from standby mode.
This difference must be taken account of in the driver.

This commit abstracts __LL_RCC_CALC_MSI_FREQ macro such that all series
can be supported, additionally the switch to the msirange
(LL_RCC_MSI_EnableRangeSelection) is now only executed on series
that support it.

As a result stm32wb socs can use msi as sysclock.
The same should be done for stm32l0, but this commit series limits
the scope of socs to avoid getting too bloated.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/clock: fix stm32 flash latency calculation in init function

This commit fixes the flash latency calulation in clock_control_init
for stm32wb and stm32wl series:
For these series new_hclk_freq can't be used to set the flash latency,
because the flash clk uses a different prescaler.

Without this fix, the flash latency could be set to an inadequat value
in cases wehere the new AHB3/AHB4 prescaler is different from the
new cpu1 prescler.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -36,6 +36,15 @@
 
+/* Calculate MSI freq for the given range(at RUN range, not after standby) */
+#if defined(CONFIG_SOC_SERIES_STM32WBX)
+#define Z_RCC_CALC_MSI_RUN_FREQ(range) __LL_RCC_CALC_MSI_FREQ( \
```
Z_ namespace is reserved for other purposes (previous `z_` macros should be fixed).

str4t0m replied:
is now called RCC_CALC_MSI_RUN_FREQ instead.

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -400,8 +409,11 @@ int stm32_clock_control_init(const struct device *dev)
 
 #if STM32_PLL_SRC_MSI
 
+
```
Nit: Extra line not needed

str4t0m replied:
removed the line.

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -47,7 +47,18 @@
+#endif
+
+#if defined(CONFIG_SOC_SERIES_STM32WBX)
+#define Z_RCC_CALC_FLASH_FREQ __LL_RCC_CALC_HCLK4_FREQ
```
Avoid `Z_` macros

str4t0m replied:
is now RCC_CALC_FLASH_FREQ instead

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -348,21 +359,28 @@ int stm32_clock_control_init(const struct device *dev)
+	flash_prescaler = s_ClkInitStruct.AHB3CLKDivider;
 #else
 	hclk_prescaler = s_ClkInitStruct.AHBCLKDivider;
+	flash_prescaler = s_ClkInitStruct.AHBCLKDivider;
```
Nit: 
For readability purpose, I'd do:
flash_prescaler = hclk_prescaler;
Just a suggestion, not blocking.

str4t0m replied:
Done.

erwango reviewed (changes requested):
Thanks for these fixes. Obviously no one tried to play much with clock settings on these series yet.
Few comments, otherwise looks good.

str4t0m committed:
drivers/clock_control: stm32: rename macros to avoid z_ namespace

A few macros used the reserved z_ prefix, this commit refactors
the driver to avoid them.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m commented:
Thanks. I have addressed the comments and added an additional commit renaming existing macros to avoid the z_ namespace.

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -19,22 +18,22 @@
 /* Macros to fill up prescaler values */
-#define z_ahb_prescaler(v) LL_RCC_SYSCLK_DIV_ ## v
-#define ahb_prescaler(v) z_ahb_prescaler(v)
+#define __ahb_prescaler(v) LL_RCC_SYSCLK_DIV_ ## v
```
Txs for taking cafe of that, but `__` is even more reserved :-)
what about `fn` ?

str4t0m replied:
Damn, this was a bit too quick. Now using fn_

FRASTM reviewed (commented):
else LGTM

str4t0m commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -56,13 +56,14 @@
  */
 static void config_bus_clk_init(LL_UTILS_ClkInitTypeDef *clk_init)
 {
+#if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_SOC_STM32WL55XX)
```
I do agree that at the current state it is superfluous and that once stm32wl54x is added it has to be extended.
I chose to do it right away with the other fixes such that it is less likely overlooked, and mainly because I'd like to push the config and a board with a stm32wle5x most likely later this week. This way I wouldn't have to take care about clock_control in the new pr.

Could you please confirm that you prefer to leave it with CONFIG_SOC_STM32WL55XX before I make any changes.
Then I'll take care of the different stm32wl socs it in the next PR. Do you think we should introduce a CONFIG_SOC_STM32WL5X and CONFIG_SOC_STM32WLEX in the future to avoid having to check for two socs sperately?

str4t0m replied:
@FRASTM I have no changed it exactly as you suggested:
Use CONFIG_SOC_SERIES_STM32WLX for now which later can be changed to CONFIG_SOC_STM32WL5X.

I do have an additional branch that introduces the remaining stm32wlxx socs on top of this and takes care about cpu2 prescaler:
https://github.com/str4t0m/zephyr/tree/stm32wl_socs
How much other stuff do you st guys have to take of in the merge window? Should I open a pr with that branch or would it be only a distraction of more important stuff?

FRASTM replied:
OK to "leave it with CONFIG_SOC_STM32WL55XX". Thanks

FRASTM replied:
I think that CONFIG_SOC_STM32WLX plus CONFIG_SOC_STM32WL5X (then another  CONFIG_SOC_STM32WLEX when device appears in zephyr) could be fine (later).
@erwango to comment ?

erwango replied:
Let's keep it that way.
For WL dual core specific code, we'll need to mimic what was done on H7 and add a `STM32WL_DUAL_CORE` symbol.

erwango replied:
EDIT: We'll need to add a new symbol. I'm not sure `STM32WL_DUAL_CORE` is the exact thing we'll need.
I could also be `STM32WL_M4_CORE`.
